sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/Expression.java

```diff
@@ -26,8 +26,15 @@
  */
 @Evolving
 public interface Expression {
+  NamedReference[] EMPTY_REFERENCE = new NamedReference[0];
+
   /**
    * Format the expression as a human readable SQL-like string.
    */
   default String describe() { return this.toString(); }
+
+  /**
+   * List of fields or columns that are referenced by this expression.
+   */
+  NamedReference[] references();
```

Comment:
Let's follow catalyst. We can provide a default implementation here, and override it in leaf expressions.

Reply:
OK